hphp/runtime/base/stat-cache.cpp

```diff
@@ -649,6 +649,13 @@ void StatCache::refresh() {
 #ifdef __linux__
   SimpleLock lock(m_lock);
 
+  // Check if we should reset the cache
+  // as part of this refresh
+  if (m_shouldClear) {
+    m_shouldClear = false;
+    reset();
```

Comment:
Are you using reset() here rather than clear() because you don't want to hit the early return on line 660 below? If we are going to do this clearing in a deferred way based on m_shouldClear (which I'm not sure is the best thing to do), then why not have this whole if be below the one on line 659-661? That way if we enter refresh() with m_shouldClear set but m_ifd == -1 we won't needlessly init the stat cache.

Reply:
Thanks, that’s the question we have as well- we were unclear if we had to do the full reset before ExecutionContext. If not, we’ll clear it inline. 

> Additionally, we're not quite sure when it's safe to reset the stat_cache, so we defer that until the requestInit function, partially because of the comment that the work in StatCache::requestInit() that must be done before ExecutionContext::requestInit. https://github.com/facebook/hhvm/blob/HHVM-3.30/hphp/runtime/base/program-functions.cpp#L2617

Reply:
There is only one stat cache that is shared by all requests, so even if you're doing it before the start of ExecutionContext::requestInit in one request, it can easily be in the middle of a request for other requests that are running concurrently.
IIUC the comment in program-functions.cpp is trying to say that we want to process any pending cache invalidations via StatCache::requestInit before we potentially start using the StatCache while initializing the rest of the request (ExecutionContext::requestInit, etc.). IIUC this ordering is needed so that, if you update a file on disk and then issue a request to the webserver, the request will definitely see that file change when it is initializing. If you're going to invalidate the entire cache by calling clear, then AFAICT it doesn't matter when it happens relative to the lifetime of the request that triggers it or other requests that are running in parallel.

Reply:
Outstanding. I’ll clean up the pr tonight to simplify and call inline. 

Thanks!